### ISSUE TITLE
Uneven CPU core utilization

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4903]
We currently start a number of threads equal to the number of CPU cores plus one (the final thread is a special utility thread). We also pin those threads to specific CPU cores.

There are a few things that are statically assigned to certain threads, and can use a lot of CPU. The most prominent one is probably the thread that's being used for a given intracluster connection. This task is often the bottleneck in clusters, and it uses only a single thread.
Similarly the btree shards for a given table are assigned to 8 randomly chosen threads.

Yet other tasks (like incoming client connections) are assigned round-robin to one of the threads.

There is no awareness in how threads are assigned as to how busy the threads already are.

Having other CPU-intensive tasks assigned to a thread that's already handling an intracluster connection for example can therefore severely limit the performance of the server.
It also makes performance somewhat unpredictable, because it will depend on which thread a given connection or btree shard ends up on.

One suggestion to avoid this:
- split the thread space up into e.g. "intracluster", "btree" and "general purpose" threads
- launch one thread per CPU core for each of those classes
- don't pin any threads to a core. Instead let the operating system dynamically rebalance the threads to whichever core is the least busy.

**Comment by danielmewes** [comment]
The fix to https://github.com/rethinkdb/rethinkdb/issues/4905#issuecomment-144866201 improved the situation a bit by trying to spread cluster connection handlers evenly over the available threads if possible.

**Comment by thedrow** [comment]
There are multiple optimizations I can come up with that will help utilize CPU more correctly. Some of them involve using the (Leader/Follower pattern)[http://kircher-schwanninger.de/michael/publications/lf.pdf] which I assume RethinkDB is already implementing at least partially since a quick search in your code came up with a threadpool, an event queue and polling mechanisms such as epoll, poll and select which are the exact building blocks for the pattern.
Static pinning does not occur in the classic implementation of the L/F pattern which may help distribute CPU more correctly but there is a proposed variant named by the paper as "Hybrid thread associations" that does just that.
The document I linked to describes the pattern quite well but I'll summarize myself for those who are not interested in reading it entirely.
In the original proposal the thread pool is divided into two categories: Leader and Followers.
The Leader is responsible for listening to I/O events and the followers are responsible for processing I/O events.
The variant I'd recommend using is to allow multiple Leaders to coexist and this is briefly mentioned in the paper.
When a Leader receives an I/O event it promotes an available follower thread to be a Leader and demotes itself to be a follower.
This allows us to scale the amount of I/O processing we are able to perform at a given moment of time by adding more Leaders or Followers accordingly.
I'm assuming that intracluster communication doesn't happen 100% of the time and that there are times where it is completely idle. One of the Leader threads (or more) can watch the intracluster file descriptor among other file handles it might handle which will allow us to process other events when that thread is idle. Furthermore, if intracluster chatter is high, more threads will be able to process those and move on to processing other events. That can happen if the order of messages coming from intracluster commnication doesn't matter.
CPU intensive tasks such as building indexes will pop the threads from the L/F pool into a dedicated pool of workers. As such I/O operations will be processed more slowly and infrequently (assuming the same rate of events is maintained) but since we're pinning threads to CPUs we can automatically scale up the L/F thread pool which is usually I/O bound so we'll be still able to at the worst case initiate the same amount of I/O events processing as before.
We can also add more Leaders based on current CPU utilization. If currently each follower thread uses 33% and the work it is performing can be paralleled we can assign two worker threads to that follower thread (new or already idle) to parallelize the work. If the work is not parallelizable we can spawn two new Leader threads that will expect that same kind of event (or alternatively same kind of CPU intensiveness) based on previous requests made to the node. 
This requires a priority queue which always attempts to keep the less CPU intensive tasks at the top of the queue (so that more work can be performed in parallel).

This way you can utilize all your CPU cores and dynamically balance between I/O and CPU bound tasks.

There are other optimization opportunities regarding b-tree indexes that may or may not depend on the architecture and hardware of your servers.
- We can distribute the building of B-Tree indexes between the available read replicas which will free the busy nodes in the cluster to respond to other events. The penalty of serialization and de-serialization of the B-Tree may or may not exceed the the value provided by this approach.
- We can optimize the building and searching of B-Trees for processors that support SIMD instructions (which are available on AWS btw) using methods like (FAST)[http://cs.nyu.edu/~lerner/spring12/Preso06-SIMDTree.pdf]. This will increase the availability of other workers to other tasks since it requires less time to build a B-Tree index this way.
- We can use a combination of SIMD instructions and GPUs to search and build the indexes as described in http://www.umiacs.umd.edu/~joseph/Indexing_IPDPS2011.pdf and http://www.mayankdaga.com/wp-content/uploads/2012/11/paper.pdf
Note: I have to read these myself, but by just skimming through them I can see that they are relevant.
- We can introduce other index types which are more efficient for memory-first databases and are suitable for hot data that is constantly being read like in http://www3.informatik.tu-muenchen.de/~leis/papers/ART.pdf

I'm not familiar with RethinkDB internals so my suggestions may be irrelevant but I'd love to hear what you guys think.

**Comment by danielmewes** [comment]
Very interesting read @thedrow . Thanks for your references and the write-up of the paper.

RethinkDB is a bit special in that it currently very heavily relies on cooperative coroutines. Many parts of the code are written to avoid the need cross-thread communication and locking, since most data structures have a "home thread". Those data structures can be used by multiple coroutines to mask i/o, but only one such coroutine will be running in a non-blocking section at a time.

We use a related mechanism to parallelize Btree operations across CPU cores. The key space is split up into (currently constant) 8 hash-based shards. Each hash-shard has its own btree structure and own page cache. The processing for each shard happens on a different thread (assuming there are at least 8 CPU cores available). However each individual shard/btree is only directly accessed by one single thread. This again avoids cross-thread locking and reduces the overhead for cross-thread communication.

There are some limitations to this approach, in that it's harder to scale to an arbitrary number of CPU cores with a single table and/or single query (the processing portion not related to btree operations for a particular query is assigned to a random thread).

Changing this architecture is a lot of work, since assumptions about the coroutine model on top of explicit thread assignments is baked pretty deeply into the code.
Unfortunately that makes it hard to do a quick comparison of what the performance differences between the different concepts will be.

Some of the other points you mention are also worth looking into once we've gotten a few lower-hanging fruits for performance optimizations out of the way. For example looking into better utilization of SIMD instructions in the btree and/or the query processing code is something we should consider once CPU profiling shows that those parts have become a bottleneck (last time I profiled RethinkDB, a bigger chunk of CPU time was associated with serialization/deserialization of queries and data, as well as with memory allocations of intermediate structures etc.).

**Comment by thedrow** [comment]
So when the dedicated intracluster thread does not perform intracluster communication it executes another task until a message is received? Or is it idle?
L/F does require coordination between threads. Running a single query in two threads also requires coordination.
I see what you're talking about. In that case launching a thread per CPU core for each class sounds like the best solution to me.
You can execute queries in parallel if you have a pool of "query" threads that any thread can make use of. This does require a bit of locking but it means that work can be parallelized from within a random thread.  Unless I am missing something of course.

Can we make the number of shards more flexible? A hashing algorithm like Murmur should allow us to add and remove shards.

**Comment by thedrow** [comment]
If memory allocation presents a problem arena allocators can be used to preallocate some of the memory required for those data types.
I quickly searched the code and it seems that this is not currently being used. If there isn't a good reason for that, lets open a new issue about it.

**Comment by thedrow** [comment]
@danielmewes [memkind](https://github.com/memkind/memkind) looks very promising for faster memory management.
It's using a customized version of jemalloc that utilizes [NUMA](https://en.wikipedia.org/wiki/Non-uniform_memory_access). It is still beta software
AMD has experimented with [a NUMA aware tcmalloc implementation](http://amd-dev.wpengine.netdna-cdn.com/wordpress/media/2012/10/NUMA_aware_heap_memory_manager_article_final.pdf).
We could use libnuma directly but it is not supported under windows.
I'd still would like to know if any of my suggestions should lead to new issues.

**Comment by danielmewes** [comment]
> So when the dedicated intracluster thread does not perform intracluster communication it executes another task until a message is received? Or is it idle?

There isn't really a dedicated intracluster thread at the moment. However we assign a given intracluster connection to a particular thread. That thread will handle other operations too.

With the proposal I mentioned in the original post, there would be dedicated intracluster threads. Those would become idle when there is no work to do on "their" set of intracluster connections. However other threads (e.g. the "btree" or "general purpose" threads) could be scheduled by the OS to run on the same core in the meantime.

> Can we make the number of shards more flexible? A hashing algorithm like Murmur should allow us to add and remove shards.

We should eventually, but it will be a major change.
It would be reasonably to use the same machinery that we use for range shards in order to balance hash shards both across different servers, as well as within a single server. We would probably combine this with a general move towards hash sharding as our only or at least primary sharding method.
At the moment we rely on the constant number of hash shards for multiplexing multiple caches / btrees into the same file, without any interference (specifically we map per-shard block IDs to per-file block IDs by using something like `file_block_id = NUM_SHARDS * shard_block_id + shard_offset`). This would have to work differently if we want to support an arbitrary number of such shards.


Regarding allocations:
We currently use jemalloc. I haven't looked at it too closely yet, but `memkind` might allow us to get some extra performance. In any case I think we should first profile RethinkDB again to find out what the actual bottlenecks now are. It has been a while since I've last done it, and some things have changed quite a bit since then.

**Comment by thedrow** [comment]
I saw that you guys use jemalloc. That's why memkind is so interesting. It makes jemalloc NUMA aware.
